A42-xds-ring-hash-lb-policy.md

```diff
@@ -361,13 +365,35 @@ The `ring_hash_experimental` policy will have the following config:
 
 ```
 message RingHashLoadBalancingConfig {
+  // A client-side option will cap these values to 4096.  If either of these
+  // values are greater than the client-side cap, they will be treated
+  // as the client-side cap value.
   uint64 min_ring_size = 1;  // Optional, defaults to 1024.
-  uint64 max_ring_size = 2;  // Optional, defaults to 8M.
+  uint64 max_ring_size = 2;  // Optional, defaults to 4096, max is 8M.
 }
 ```
 
 These parameters are used to determine how the ring is constructed, just
-as they are in the Envoy implementation.
+as they are in the Envoy implementation.  However, in order to limit the
+possibility of a control plane causing a client to OOM by creating a lot
+of large rings, we want to limit the ring size to a much smaller value
+than Envoy does.  The `max_ring_size` field will default to 4096 instead
+of 8M as in Envoy, although it will still accept values up to 8M for
```

Comment:
I'm not suggesting that we add this; this is the current behavior.  We currently NACK if the value is greater than 8M.  That's the same behavior as Envoy.

Reply:
(Some of this is related to that other thread, but talking here to avoid splitting the conversation.)

In Java we had a 8M cap, but we didn't NACK when it was exceeded. If you want to tell the control plane you aren't following it, a NACK makes sense. But that also makes it a PITA if the limit ever changes as you don't know a priori the xds client's limit.

I'm not too concerned one way or another[1], but the "accept up to 8M" wouldn't have been clearly "we NACK when over 8M". So if you want that, probably say that explicitly.

1. I think clamping is the easier thing to do in a practical sense for control planes. Certainly NACKing at 4 KiB would be annoying to the ecosystem. I can see some argument for NACKing above 8M, but I personally feel it is a bit misguided. But none of that really matters and I'm fine either way if that was the intention and had been overlooked before.

Reply:
Thanks for the discussion.  My default assumption is that we should NACK the same conditions that Envoy does, unless we have a compelling reason to do otherwise.  I've updated the wording here to explicitly say that we NACK values greater than 8M.